src/core/datatypes/trees/Tree.cpp

```diff
@@ -1585,6 +1587,48 @@ void Tree::collapseSampledAncestors()
     setRoot(root, true);
 }
 
+// Pick a random node which is not the root, a tip, or the parent of a sampled ancestor.
+// First try doing so at random; if that does not work, use the more computationally demanding strategy of finding all eligible nodes.
+TopologyNode* Tree::pickRandomInternalNode(RandomNumberGenerator* rng) const
+{
+    TopologyNode* node;
+    int counter = 0;
+    
+    do {
+        double u = rng->uniform01();
+        size_t node_idx = size_t( std::floor(getNumberOfNodes() * u) );
+        node = (TopologyNode*)&getNode(node_idx);
+        counter++;
```

Comment:
We could simplify this further by just making it a for loop that counts to 10, and then after we sample a node we just `return` immediately if it satisfies the condition:

```
if ( not (node->isRoot() || node->isTip() || node->isSampledAncestorParent() ))) return node;
```

Then we could also remove the `if (counter == 10)` condition and unnest the code inside it.